File: models/video.go
package models

import (
	"time"

	"github.com/lib/pq"
	"gorm.io/gorm"
)

//Episode 分集
type Episode struct {
	gorm.Model
	VideoID   int64
	Num       float64
	Title     string `gorm:"size:50;NOT NULL;"`
	Desc      string `gorm:"size:5000;NOT NULL;"`
	Cover     string `gorm:"size:500;NOT NULL;"`
	URL       string `gorm:"size:500;NOT NULL;"`
	MobileURL string `gorm:"size:500;"`
	Subtitles []*Subtitle
}

//Subtitle 字幕
type Subtitle struct {
	gorm.Model
	EpisodeID uint
	Name      string `gorm:"size:50;NOT NULL;"`
	URL       string `gorm:"size:500;NOT NULL;"`
}

//Video 视频
type Video struct {
	gorm.Model
	Title    string `gorm:"size:50;NOT NULL;"`
	Desc     string `gorm:"size:5000;NOT NULL;"`
	PubDate  time.Time
	Cover    string `gorm:"size:500;NOT NULL;"`
	Episodes []*Episode
	Tags     pq.StringArray `gorm:"type:varchar(10)[]"`
	IsShow   bool
}

//VideoSeriesItem 视频系列视频列表
type VideoSeriesItem struct {
	VideoSeriesID uint   `gorm:"primary_key;auto_increment:false"`
	VideoID       uint   `gorm:"primary_key;auto_increment:false"`
	Title         string `gorm:"-"`
	Alias         string
	Num           int
}

//VideoSeries 视频系列
type VideoSeries struct {
	gorm.Model
	Name  string
	Items []*VideoSeriesItem
}

File: web/src/models/video.ts

export interface Subtitle {
    name: string
    url: string
}

export interface Episode {
    id: number
    videoID: number
    title: string
    desc: string
    num: number
    cover: string
    url: string
    subtitles: Subtitle[]
}

export interface VideoCardModel {
    id: number
    title: string
    desc: string
    cover: string
    totalNum: number
    episodeID: number
}

export interface IUpdateVideo {
    title: string,
    desc: string,
    cover: string,
    pubDate: number,
    tags: string[],
    isShow: boolean,
    isHideOnMobile: boolean,
    theme: string
}

export const getVideoDetail = (data: any, episodeID: string | null) => {
    const videos = data?.videos.edges
    let video: any
    if (videos?.length > 0) {
        video = videos[0]
    }
    let videoItem = {
        id: 0,
        cover: "",
        title: "",
        desc: "",
        tags: [],
        pubDate: 0,
        episodes: [],
        theme: ""
    }
    if (video) {
        videoItem = {
            id: video.id,
            cover: video.cover,
            title: video.title,
            desc: video.desc,
            episodes: video.episodes,
            theme: video.theme,
            tags: video.tags,
            pubDate: video.pubDate,
        }
    }
    let num = 0
    if (video && video.episodes) {
        for (let i = 0, len = video.episodes.length; i < len; i++) {
            if (Number(video.episodes[i].id) === Number(episodeID)) {
                num = i
                break
            }
        }
    }
    let episodeItem = {
        id: 0,
        url: "",
        subtitles: null
    }
    if (video && video.episodes && num < video.episodes.length) {
        episodeItem = {
            id: video.episodes[num].id,
            url: video.episodes[num].url,
            subtitles: video.episodes[num].subtitles,
        }
    }
    return {
        video: video,
        videoItem: videoItem,
        num: num,
        episodeItem: episodeItem,
        videoSerieses: data?.videoSerieses.edges,
        similarVideos: data?.similarVideos.edges
    }
}